pkg/buildpack/module_manager_test.go

```diff
@@ -0,0 +1,299 @@
+
+		compositeBP1, err = ifakes.NewFakeBuildpack(dist.BuildpackDescriptor{
+			WithAPI: api.MustParse("0.2"),
+			WithInfo: dist.ModuleInfo{
+				ID:      "composite-buildpack-1-id",
+				Version: "composite-buildpack-1-version",
+			},
+			WithOrder: []dist.OrderEntry{{
+				Group: []dist.ModuleRef{
+					{
+						ModuleInfo: bp1.Descriptor().Info(),
+					},
+					{
+						ModuleInfo: compositeBP2.Descriptor().Info(),
+					},
+				},
+			}},
+		}, 0644)
+		h.AssertNil(t, err)
+	})
+
+	when("manager is configured in flatten mode", func() {
+		when("flatten all", func() {
+			it.Before(func() {
+				moduleManager = buildpack.NewModuleManager(true, buildpack.FlattenMaxDepth)
+				moduleManager.AddModules(compositeBP1, []buildpack.BuildModule{bp1, compositeBP2, bp21, bp22, compositeBP3, bp31}...)
+			})
+
+			when("#FlattenModules", func() {
+				it("returns one flatten module (1 layer)", func() {
+					modules := moduleManager.FlattenModules()
+					h.AssertEq(t, len(modules), 1)
+					h.AssertEq(t, len(modules[0]), 7)
+				})
+			})
+
+			when("#NonFlattenModules", func() {
+				it("returns empty", func() {
+					modules := moduleManager.NonFlattenModules()
+					h.AssertEq(t, len(modules), 0)
+				})
+			})
+
+			when("#AllModules", func() {
+				it("returns all modules", func() {
+					modules := moduleManager.AllModules()
+					h.AssertEq(t, len(modules), 7)
+				})
+			})
+
+			when("#IsFlatten", func() {
+				it("returns true for flatten modules", func() {
+					h.AssertTrue(t, moduleManager.IsFlatten(compositeBP1))
+					h.AssertTrue(t, moduleManager.IsFlatten(bp1))
+					h.AssertTrue(t, moduleManager.IsFlatten(compositeBP2))
+					h.AssertTrue(t, moduleManager.IsFlatten(bp21))
+					h.AssertTrue(t, moduleManager.IsFlatten(bp22))
+					h.AssertTrue(t, moduleManager.IsFlatten(compositeBP3))
+					h.AssertTrue(t, moduleManager.IsFlatten(bp31))
+				})
+			})
+		})
+
+		when("flatten with depth=1", func() {
+			it.Before(func() {
+				moduleManager = buildpack.NewModuleManager(true, 1)
+				moduleManager.AddModules(compositeBP1, []buildpack.BuildModule{bp1, compositeBP2, bp21, bp22, compositeBP3, bp31}...)
+			})
+
+			when("#FlattenModules", func() {
+				it("returns 1 flatten module with [compositeBP2, bp21, bp22, compositeBP3, bp31]", func() {
+					modules := moduleManager.FlattenModules()
+					h.AssertEq(t, len(modules), 1)
+					h.AssertEq(t, len(modules[0]), 5)
+				})
+			})
+
+			when("#IsFlatten", func() {
+				it("returns true for flatten modules", func() {
+					h.AssertTrue(t, moduleManager.IsFlatten(compositeBP2))
+					h.AssertTrue(t, moduleManager.IsFlatten(bp21))
+					h.AssertTrue(t, moduleManager.IsFlatten(bp22))
+					h.AssertTrue(t, moduleManager.IsFlatten(compositeBP3))
+					h.AssertTrue(t, moduleManager.IsFlatten(bp31))
+				})
+
+				it("returns false for no flatten modules", func() {
+					h.AssertFalse(t, moduleManager.IsFlatten(bp1))
+					h.AssertFalse(t, moduleManager.IsFlatten(compositeBP1))
```

Comment:
I have to confess... I'm still not quite sure what `IsFlatten` is supposed to mean...

Reply:
well, the idea is this: I will add Buildpacks to the *moduleManager* and the manager must take care of organizing those Buildpacks in two groups (according to the configuration provided): 
- Buildpacks that must be treated as individuals
- Buidlapcks that must be flattened

*IsFlatten* is supposed to answer the question: *Should this given Buildpack be treated as a flattened one?*

Reply:
Ah... in this case, maybe a better name is `ShouldFlatten`?